### ISSUE TITLE
Segfault on setting gradient value to instance of user-defined class

### ISSUE THREAD

**Issue by knottb** [pytorch/pytorch#64813]
## 🐛 Bug

Seems to be a segfault when setting the value of a parameter.grad to a user-defined class instance. This is probably not useful, but I wanted to document it. It would be better for this crash to raise an error rather than issue a segfault if possible.

## To Reproduce

Minimal reproduction code:

```
class Example:
    def __init__(self):
        self.x = 0
        
x = torch.tensor([1., 2., 3., 4., 5.])
model = torch.nn.Linear(5, 1)
model(x).backward()

model.weight.grad = Example()
```

## Expected behavior

I imagine there could be some expected behavior, but this should probably remain undefined, but should be gracefully handled by raising an exception rather than a segfault.

## Environment

PyTorch Version: 1.10.0.dev20210831

## Additional context

I came across this during a workflow where I set `param.grad = crypten.cryptensor(param.grad)` and manipulate it in CrypTen. Instead, I have to use a new variable then decrypt before setting value into `param.grad`. This is fine, however during testing, my multiprocess tests did not catch the issue even with this failure due to the segfault. Had this raised an exception, my test would have recognized that this shouldn't pass. 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by Varal7** [comment]
This was introduced here: https://github.com/pytorch/pytorch/pull/56017/files#diff-0fa862a86ab690b67dbc2aecb5cc0675d6df08af87e8a2b0655e554cef6f11d7L438.

When the check 
```c++
THPUtils_assertRet(-1, THPVariable_Check(py_grad),
       "expected Variable or None (got %s)", THPUtils_typename(py_grad));
```

was removed from `python_variable.cpp` on line 438 of the linked PR.
